src/app/server/Server.cpp

```diff
@@ -70,6 +73,115 @@ constexpr bool useTestPairing()
 #endif
 }
 
+class ServerStorageDelegate : public PersistentStorageDelegate
+{
+    void SetDelegate(PersistentStorageResultDelegate * delegate) override {}
+    void GetKeyValue(const char * key) override {}
+    void SetKeyValue(const char * key, const char * value) override {}
+
+    CHIP_ERROR GetKeyValue(const char * key, void * buffer, uint16_t & size) override
+    {
+        return PersistedStorage::KeyValueStoreMgr().Get(key, buffer, size);
+    }
+
+    CHIP_ERROR SetKeyValue(const char * key, const void * value, uint16_t size) override
+    {
+        return PersistedStorage::KeyValueStoreMgr().Put(key, value, size);
+    }
+
+    void DeleteKeyValue(const char * key) override { PersistedStorage::KeyValueStoreMgr().Delete(key); }
+};
+
+ServerStorageDelegate gServerStorage;
+
+CHIP_ERROR PersistAdminPairingToKVS(AdminPairingInfo * admin, AdminId nextAvailableId)
+{
+    ReturnErrorCodeIf(admin == nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    ChipLogProgress(AppServer, "Persisting admin ID %d, next available %d", admin->GetAdminId(), nextAvailableId);
+
+    ReturnErrorOnFailure(admin->StoreIntoKVS(gServerStorage));
+    ReturnErrorOnFailure(PersistedStorage::KeyValueStoreMgr().Put(kAdminTableCountKey, &nextAvailableId, sizeof(nextAvailableId)));
+
+    ChipLogProgress(AppServer, "Persisting admin ID successfully");
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR RestoreAllAdminPairingsFromKVS(AdminPairingTable & adminPairings, AdminId & nextAvailableId)
+{
+    // It's not an error if the key doesn't exist. Just return right away.
+    VerifyOrReturnError(PersistedStorage::KeyValueStoreMgr().Get(kAdminTableCountKey, &nextAvailableId) == CHIP_NO_ERROR,
+                        CHIP_NO_ERROR);
+    ChipLogProgress(AppServer, "Next available admin ID is %d", nextAvailableId);
+
+    // TODO: The admin ID space allocation should be re-evaluated. With the current approach, the space could be
+    //       exhausted while IDs are still available (e.g. if the admin IDs are allocated and freed over a period of time).
+    //       Also, the current approach can make ID lookup slower as more IDs are allocated and freed.
+    for (AdminId id = 0; id < nextAvailableId; id++)
+    {
+        AdminPairingInfo * admin = adminPairings.AssignAdminId(id);
+        // Recreate the binding if one exists in persistent storage. Else skip to the next ID
+        if (admin->FetchFromKVS(gServerStorage) != CHIP_NO_ERROR)
+        {
+            adminPairings.ReleaseAdminId(id);
+        }
+        else
+        {
+            ChipLogProgress(AppServer, "Found admin pairing for %d, node ID %llu", admin->GetAdminId(), admin->GetNodeId());
+        }
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void DeleteAdminPairingsInKVS(AdminId nextAvailableId)
```

Comment:
Could we rename these a bit?
As a reader I would consider `DeleteAdminPairingsInKVS(id)` to mean 'delete the pairings for the specified id from KVS' however this does a full clean. How about `EraseAllAdminPairingsUpTo(id)` and that may be clearer